File: crates/pray-core/src/render_patch.rs
use crate::hashing::checksum_managed_body_line_refs;
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Eq)]
enum Segment {
    Text(String),
    Managed { id: String, body: String },
}

/// Preserve unmarked text from `existing` while replacing managed spans from `fresh`.
///
/// When `existing` has no managed spans that overlap `fresh`, return `fresh` wholesale
/// so corrupted or empty destinations can be repaired by a full rewrite.
/// When `fresh` introduces managed spans before the first shared id, take that
/// leading region from `fresh` so local compose sources can gain markers and update.
pub fn patch_rendered_content(existing: &str, fresh: &str) -> String {
    let existing_segments = split_segments(existing);
    let fresh_segments = split_segments(fresh);
    let fresh_managed: BTreeMap<String, String> = fresh_segments
        .iter()
        .filter_map(|segment| match segment {
            Segment::Managed { id, body } => Some((id.clone(), body.clone())),
            Segment::Text(_) => None,
        })
        .collect();
    let existing_ids: std::collections::BTreeSet<&str> = existing_segments
        .iter()
        .filter_map(|segment| match segment {
            Segment::Managed { id, .. } => Some(id.as_str()),
            Segment::Text(_) => None,
        })
        .collect();
    let existing_overlap = existing_segments.iter().any(|segment| match segment {
        Segment::Managed { id, .. } => fresh_managed.contains_key(id),
        Segment::Text(_) => false,
    });
    if !existing_overlap {
        return fresh.to_string();
    }
    let mut used = std::collections::BTreeSet::new();
    let mut output = String::new();
    let remaining_existing = if fresh_managed
        .keys()
        .any(|id| !existing_ids.contains(id.as_str()))
    {
        let Some(shared_id) = fresh_segments.iter().find_map(|segment| match segment {
            Segment::Managed { id, .. } if existing_ids.contains(id.as_str()) => Some(id.as_str()),
            _ => None,
        }) else {
            return fresh.to_string();
        };
        for segment in &fresh_segments {
            match segment {
                Segment::Managed { id, .. } if id == shared_id => break,
                Segment::Text(text) => output.push_str(text),
                Segment::Managed { id, body } => {
                    used.insert(id.clone());
                    push_managed_span(&mut output, id, body);
                }
            }
        }
        skip_until_managed(&existing_segments, shared_id)
    } else {
        existing_segments.as_slice()
    };
    for (index, segment) in remaining_existing.iter().enumerate() {
        match segment {
            Segment::Text(text) => output.push_str(text),
            Segment::Managed { id, body } => {
                let replacement = fresh_managed
                    .get(id)
                    .cloned()
                    .unwrap_or_else(|| body.clone());
                used.insert(id.clone());
                push_managed_span(&mut output, id, &replacement);
                if let Some(Segment::Managed { id: next_id, .. }) =
                    remaining_existing.get(index + 1)
                {
                    output.push_str(&text_between(&fresh_segments, id, next_id));
                }
            }
        }
    }
    append_unused_fresh_spans(&fresh_segments, &used, &mut output);
    if !output.ends_with('\n') {
        output.push('\n');
    }
    output
}

fn text_between(segments: &[Segment], left: &str, right: &str) -> String {
    let mut copying = false;
    let mut text = String::new();
    for segment in segments {
        match segment {
            Segment::Managed { id, .. } if id == left => {
                copying = true;
                text.clear();
            }
            Segment::Managed { id, .. } if copying => {
                return if id == right { text } else { String::new() };
            }
            Segment::Text(body) if copying => text.push_str(body),
            _ => {}
        }
    }
    String::new()
}

fn append_unused_fresh_spans(
    fresh_segments: &[Segment],
    used: &std::collections::BTreeSet<String>,
    output: &mut String,
) {
    let mut pending_text = String::new();
    let mut seen_used = false;
    for segment in fresh_segments {
        match segment {
            Segment::Managed { id, body } => {
                if used.contains(id) {
                    seen_used = true;
                    pending_text.clear();
                    continue;
                }
                output.push_str(&pending_text);
                pending_text.clear();
                push_managed_span(output, id, body);
            }
            Segment::Text(text) => {
                if seen_used {
                    pending_text.push_str(text);
                }
            }
        }
    }
}

fn skip_until_managed<'a>(segments: &'a [Segment], id: &str) -> &'a [Segment] {
    match segments.iter().position(|segment| match segment {
        Segment::Managed { id: found, .. } => found == id,
        Segment::Text(_) => false,
    }) {
        Some(index) => &segments[index..],
        None => segments,
    }
}

fn push_managed_span(output: &mut String, id: &str, body: &str) {
    output.push_str(&format!("<!-- pray:{id} -->\n"));
    if !body.is_empty() {
        output.push_str(body.trim_end_matches('\n'));
        output.push('\n');
    }
    output.push_str(&format!("<!-- pray:{id} -->\n"));
}

fn split_segments(content: &str) -> Vec<Segment> {
    let lines: Vec<&str> = content.lines().collect();
    let mut segments = Vec::new();
    let mut text = String::new();
    let mut index = 0usize;
    while index < lines.len() {
        if let Some(id) = marker_id(lines[index]) {
            if let Some(close) = find_closing_marker(&lines, index + 1, &id) {
                if !text.is_empty() {
                    segments.push(Segment::Text(std::mem::take(&mut text)));
                }
                let body_lines = &lines[index + 1..close];
                let body = if body_lines.is_empty() {
                    String::new()
                } else {
                    let mut body = body_lines.join("\n");
                    body.push('\n');
                    body
                };
                let _ = checksum_managed_body_line_refs(body_lines);
                segments.push(Segment::Managed { id, body });
                index = close + 1;
                continue;
            }
        }
        text.push_str(lines[index]);
        text.push('\n');
        index += 1;
    }
    if !text.is_empty() {
        segments.push(Segment::Text(text));
    }
    segments
}

fn find_closing_marker(lines: &[&str], start: usize, id: &str) -> Option<usize> {
    lines
        .iter()
        .enumerate()
        .skip(start)
        .find_map(|(index, line)| (marker_id(line).as_deref() == Some(id)).then_some(index))
}

fn marker_id(line: &str) -> Option<String> {
    let trimmed = line.trim();
    let remainder = trimmed.strip_prefix("<!-- pray:")?;
    let id = remainder.strip_suffix(" -->")?;
    if id == "0 ignore-comments" {
        return None;
    }
    Some(id.to_string())
}

File: rubygems/pray-cli/lib/pray/render_patch.rb
# frozen_string_literal: true

module Pray
  module RenderPatch
    module_function

    def patch_rendered_content(existing, fresh)
      existing_segments = split_segments(existing)
      fresh_segments = split_segments(fresh)
      fresh_managed = managed_bodies(fresh_segments)
      return fresh unless overlapping_managed?(existing_segments, fresh_managed)

      used = {}
      output = +""
      remaining = prepend_new_leading_spans(
        existing_segments,
        fresh_segments,
        managed_id_set(existing_segments),
        used,
        output
      )
      splice_existing_managed(remaining, fresh_segments, fresh_managed, used, output)
      append_unused_fresh_managed(fresh_segments, used, output)
      output.end_with?("\n") ? output : "#{output}\n"
    end

    def managed_bodies(segments)
      segments.filter_map { |segment| [segment[:id], segment[:body]] if segment[:kind] == :managed }.to_h
    end

    def overlapping_managed?(existing_segments, fresh_managed)
      existing_segments.any? { |segment| segment[:kind] == :managed && fresh_managed.key?(segment[:id]) }
    end

    def managed_id_set(segments)
      segments.filter_map { |segment| segment[:id] if segment[:kind] == :managed }.to_h { |id| [id, true] }
    end

    def splice_existing_managed(remaining, fresh_segments, fresh_managed, used, output)
      remaining.each_with_index do |segment, index|
        if segment[:kind] == :text
          output << segment[:text]
          next
        end

        used[segment[:id]] = true
        output << managed_segment(segment[:id], fresh_managed.fetch(segment[:id], segment[:body]))
        next_segment = remaining[index + 1]
        next unless next_segment && next_segment[:kind] == :managed

        output << text_between(fresh_segments, segment[:id], next_segment[:id])
      end
    end

    def text_between(segments, left, right)
      copying = false
      text = +""
      segments.each do |segment|
        if segment[:kind] == :managed
          if segment[:id] == left
            copying = true
            text = +""
          elsif copying
            return (segment[:id] == right) ? text : ""
          end
        elsif copying
          text << segment[:text]
        end
      end
      ""
    end

    def append_unused_fresh_managed(fresh_segments, used, output)
      pending = +""
      seen_used = false
      fresh_segments.each do |segment|
        if segment[:kind] == :managed
          if used[segment[:id]]
            seen_used = true
            pending = +""
            next
          end
          output << pending
          pending = +""
          output << managed_segment(segment[:id], segment[:body])
        elsif seen_used
          pending << segment[:text]
        end
      end
    end

    def prepend_new_leading_spans(existing_segments, fresh_segments, existing_ids, used, output)
      missing = fresh_segments.any? { |segment| segment[:kind] == :managed && !existing_ids[segment[:id]] }
      return existing_segments unless missing

      shared_id = fresh_segments.find { |segment| segment[:kind] == :managed && existing_ids[segment[:id]] }&.[](:id)
      return existing_segments unless shared_id

      fresh_segments.each do |segment|
        break if segment[:kind] == :managed && segment[:id] == shared_id

        if segment[:kind] == :text
          output << segment[:text]
        else
          used[segment[:id]] = true
          output << managed_segment(segment[:id], segment[:body])
        end
      end
      skip_until_managed(existing_segments, shared_id)
    end

    def relocate_managed_spans(content, spans)
      positions = marker_positions(lines_of(content))
      spans.map do |span|
        position = positions[span.id]
        next span unless position

        span.dup.tap do |relocated|
          relocated.open_line = position[0]
          relocated.close_line = position[1]
        end
      end
    end

    def split_segments(content)
      lines = lines_of(content)
      segments = []
      text = +""
      index = 0
      while index < lines.length
        identifier = marker_id(lines[index])
        close = find_closing_marker(lines, index + 1, identifier) if identifier
        if identifier && close
          segments << {kind: :text, text: text} unless text.empty?
          text = +""
          body_lines = lines[(index + 1)...close]
          body = body_lines.empty? ? "" : "#{body_lines.join("\n")}\n"
          segments << {kind: :managed, id: identifier, body: body}
          index = close + 1
        else
          text << "#{lines[index]}\n"
          index += 1
        end
      end
      segments << {kind: :text, text: text} unless text.empty?
      segments
    end

    def skip_until_managed(segments, identifier)
      index = segments.index { |segment| segment[:kind] == :managed && segment[:id] == identifier }
      index ? segments[index..] : segments
    end

    def lines_of(content)
      content.lines(chomp: true)
    end

    def find_closing_marker(lines, start, identifier)
      (start...lines.length).find { |index| marker_id(lines[index]) == identifier }
    end

    def marker_positions(lines)
      positions = {}
      active = nil
      lines.each_with_index do |line, index|
        identifier = marker_id(line)
        next unless identifier&.match?(/\A[a-z0-9]+\z/)

        if active.nil?
          active = [identifier, index + 1]
        elsif active[0] == identifier
          positions[identifier] = [active[1], index + 1]
          active = nil
        end
      end
      positions
    end

    def marker_id(line)
      match = line.strip.match(/\A<!-- pray:(.+) -->\z/)
      identifier = match && match[1]
      identifier unless identifier == "0 ignore-comments"
    end

    def managed_segment(identifier, body)
      content = body.empty? ? "" : "#{body.sub(/\n+\z/, "")}\n"
      "<!-- pray:#{identifier} -->\n#{content}<!-- pray:#{identifier} -->\n"
    end
  end
end
